### ISSUE TITLE
Streaming JsonSerializer buffers all content of skipped JSON properties.

### ISSUE THREAD

**Issue by krwq** [dotnet/runtime#96559]
When reading large JSON files which don't have fully defined model serializer may OOM instead of skipping through unused stuff.

Here is the code which generates repro, also showed how defining model is not causing OOM:

```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

string fileName = "foo.json";
const long minFileSize = 5L * 1024 * 1024 * 1024; // 5GB

if (File.Exists(fileName))
{
    using (FileStream fs = new FileStream(fileName, FileMode.Open))
    {
        Repro(fs);
    }
}
else
{
    using (FileStream fs = new FileStream(fileName, FileMode.CreateNew))
    {
        Console.Write("Generating large JSON file... ");
        GenerateLargeJson(fs, minFileSize); // increase as needed
        Console.WriteLine("Done");
        fs.Position = 0;
        Repro(fs);
    }
}

void Repro(FileStream fs)
{
    Console.Write("Deserializing (expected to work)... ");
    JsonSerializer.Deserialize<FooBarClass>(fs);
    Console.WriteLine("Done");

    fs.Position = 0;

    Console.Write("Deserializing (repro)... ");
    JsonSerializer.Deserialize<FooClass>(fs);
    Console.WriteLine("Done"); // <-- never reached, prints "Out of memory."
}


void GenerateLargeJson(FileStream fs, long minSize)
{
    Utf8JsonWriter writer = new Utf8JsonWriter(fs);
    writer.WriteStartObject(); // root

    writer.WriteString("Foo", "FooValue");

    writer.WritePropertyName("Bars");
    writer.WriteStartObject();

    long propIdx = 0;
    do
    {
        for (int i = 0; i < 1000; i++, propIdx++)
        {
            writer.WritePropertyName($"PropertyWithSomeQuiteLongishName{propIdx:000000000}");
            writer.WriteStartObject();

            writer.WriteString("Bar", $"BarValue{propIdx:000000000}SomethingNotTooShortSoWeDontNeedToWriteThatManyTimes");

            writer.WriteEndObject();
        }

        writer.Flush();

    } while (fs.Length < minSize);

    writer.WriteEndObject(); // Bars

    writer.WriteEndObject(); // root
    writer.Flush();
}

class FooClass
{
    public string Foo { get; set; }
}

// Not needed for repro:
class FooBarClass
{
    public string Foo { get; set; }
    public Dictionary<string, BarClass> Bars { get; set; }
}

class BarClass
{
    public string Bar { get; set; }
}
```

**Comment by krwq** [comment]
I've just run DeserializeAsync to confirm that is also affected and the results are identical.

**Comment by eiriktsarpalis** [comment]
Root cause appears to be this logic in the async object converter:

https://github.com/dotnet/runtime/blob/710774a3117dbf62ab35565f4b0b300f7a775319/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs#L195-L206

TL;DR the `TrySkip()` method designed for partial JSON payloads still needs to buffer the full payload to perform the final skip which results in the OOM exception. We would need to add a new method on `Utf8JsonReader` that supports partial skipping.

**Comment by gregsdennis** [comment]
@eiriktsarpalis with this kind of parsing, is the stream lost or can it be recovered?  For instance, if I wanted to go back and read different portions of the same content in multiple operations, can I do that?

**Comment by eiriktsarpalis** [comment]
You lose everything apart from the window captured by the current `Utf8JsonReader` (which can be snapshotted cheaply by copying the struct). This isn't really a concern when writing custom converters, as the serializer will ensure that the current JSON value is fully buffered before calling the your converter. You're out of luck if you wanted to rewind back to previous values, streaming JSON serialization doesn't really cater to that use case.
